reactor-extra/src/main/java/reactor/retry/RetryBuilder.java

```diff
@@ -0,0 +1,222 @@
+import java.util.function.Predicate;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.EmitterProcessor;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+
+/**
+ * Retry utility to control retry or repeat of {@link Flux} or @link {@link Mono} using
+ * <ul>
+ *   <li>{@link Flux#retryWhen(Function)} </li>
+ *   <li>{@link Flux#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#retryWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(int, Function)} </li>
+ * </ul>
+ *
+ * This class enables commonly used retry/repeat patterns:
+ * <ul>
+ *   <li>Configured number of retries without backoff</li>
+ *   <li>Configured number of retries with fixed backoff interval</li>
+ *   <li>Exponential backoff with timeout</li>
+ *   <li>Retriable exception configuration that can be used with/without backoff</li>
+ *   <li>Non-retriable exception configuration that can be used with/without backoff</li>
+ *   <li>doOnRetry to perform any rollback operation before performing retry</li>
+ * </ul>
+ *
+ * This class does not support fully flexible functions which can be done just as easily on {@link Flux#retryWhen(Function)} etc.
+ * <ul>
+ *   <li>Generic predicate for retry decision</li>
+ *   <li>Flexible delay based on context</li>
+ * </ul>
+ *
+ */
+public class RetryBuilder {
+
+	static final Duration NO_RETRY = Duration.ofSeconds(-1);
+
+	final Predicate<Throwable> retryPredicate;
+	Integer maxAttempts;
+	Duration timeout;
+	Instant timeoutInstant;
+	Function<Long, Duration> backoffCalculator;
+	Scheduler backoffScheduler;
+	Consumer<RetryContext> onRetry;
+
+
+	RetryBuilder(Predicate<Throwable> retryPredicate) {
+		this.retryPredicate = retryPredicate;
+		this.maxAttempts = null;
+		this.backoffCalculator = iteration -> Duration.ZERO;
+		this.onRetry = r -> {};
+	}
+
+	public static RetryBuilder any() {
+		return new RetryBuilder(e -> true);
+	}
+
+	@SafeVarargs
+	public static RetryBuilder anyOf(Class<? extends Throwable>... retriableExceptions) {
+		return new RetryBuilder(e -> {
+			for (Class<? extends Throwable> clazz : retriableExceptions) {
+				if (clazz.isInstance(e))
+					return true;
+			}
+			return false;
+		});
+	}
+
+	@SafeVarargs
+	public static RetryBuilder allBut(final Class<? extends Throwable>... nonRetriableExceptions) {
+		return new RetryBuilder(e -> {
+			for (Class<? extends Throwable> clazz : nonRetriableExceptions) {
+				if (clazz.isInstance(e))
+					return false;
+			}
+			return true;
+		});
+	}
+
+	public RetryBuilder once() {
+		this.maxAttempts = 1;
+		return this;
+	}
+
+	public RetryBuilder maxAttempts(int maxAttempts) {
+		this.maxAttempts = maxAttempts;
+		return this;
+	}
+
+	public RetryBuilder timeout(Duration timeout) {
+		this.timeout = timeout;
+		return this;
+	}
+
+	public RetryBuilder noBackoff() {
```

Comment:
Backoff used at scale can impact any remote systems in a negative way. 

In a scenario where a single service is disconnected by a network partition, all clients on the other side of the partition are disconnected at the same time. If the partition ends, backoff very likely yields the same result for all clients and all clients retry at the same time which generates a fair amount of concurrent load. 

How about including a decorrelated jitter backoff strategy for such cases? (see https://www.awsarchitectureblog.com/2015/03/backoff.html for details and [DecorrelatedJitterDelay.java](https://github.com/lettuce-io/lettuce-core/blob/master/src/main/java/com/lambdaworks/redis/resource/DecorrelatedJitterDelay.java) for a sample implementation).